atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,146 @@
+    private final Disposable refreshableSubscriptionDisposable;
+    private final ReadWriteLock isClosedLock;
+
+    @GuardedBy("isClosedLock")
+    private volatile boolean isClosed;
+
+    private <K> ReloadingCloseableContainer(Refreshable<K> refreshableFactoryArgument, Function<K, T> factory) {
+        this.isClosed = false;
+        this.isClosedLock = new ReentrantReadWriteLock(true);
+        this.currentResource = new AtomicReference<>(Optional.empty());
+        this.refreshableResource = refreshableFactoryArgument.map(factoryArg -> createNewResource(factoryArg, factory));
+
+        this.refreshableSubscriptionDisposable = refreshableResource.subscribe(resource -> {
+            Optional<T> maybeResourceToClose = currentResource.getAndSet(Optional.of(resource));
+            maybeResourceToClose.ifPresent(this::shutdownResource);
+        });
+    }
+
+    public static <T extends AutoCloseable, K> ReloadingCloseableContainer<T> of(
+            Refreshable<K> refreshableFactoryArgument, Function<K, T> factory) {
+        return new ReloadingCloseableContainer<>(refreshableFactoryArgument, factory);
+    }
+
+    /**
+     * isClosedLock: See {@link #close()}.
+     */
+    private <K> T createNewResource(K factoryArg, Function<K, T> factory) {
+        return runIfNotClosed(
+                () -> factory.apply(factoryArg), "Attempted to create a new resource after the container was closed.");
+    }
+
+    /**
+     * A lock is taken out to ensure no new resources are created after retrieving the current stored
+     * resource to close. By doing so, we avoid closing a cluster and subsequently creating a new one that is
+     * never closed.
+     */
+    @Override
+    public void close() {
+        runWithWriteLock(() -> {
+            isClosed = true;
+            refreshableSubscriptionDisposable.dispose();
+            currentResource.get().ifPresent(this::shutdownResource);
+        });
+    }
+
+    /**
+     * Gets the latest resource that reflects any changes in the refreshable, provided {@link #close()} has not
+     * been called.
+     *
+     * The resource returned will be closed after {@link #close} is called, or the refreshable is refreshed, even
+     * if the resource is in active use.
+     *
+     * @throws SafeIllegalStateException if the container was closed prior to getting a resource.
+     */
+    public T get() {
+        return runIfNotClosed(refreshableResource, "Attempted to get a resource after the container was closed");
+    }
+
+    private <K> K runIfNotClosed(Supplier<K> supplier, @CompileTimeConstant String ifClosedExceptionMessage) {
+        if (!isClosed) {
+            return runWithReadLock(() -> {
+                if (isClosed) {
+                    throw new SafeIllegalStateException(ifClosedExceptionMessage);
+                }
+                return supplier.get();
+            });
+        }
+        throw new SafeIllegalStateException(ifClosedExceptionMessage);
```

Comment:
personal preference (others may disagree): for a "simple" if-then-else, can you wrap this part in `else`? IMO, it makes these kinds of things clearer (where it may look weird for a method to just plainly end in "throw new ...Exception").

Reply:
I tend to disagree with this, but agree when it's if () do work with return, or throw, without the else it's weird.
Could be clearer with standard guard statement if(isClosed) throw. Do work